zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -982,6 +991,75 @@ where
         }
         .boxed()
     }
+
+    fn z_list_unified_receivers(
+        &self,
+        address: String,
+    ) -> BoxFuture<Result<unified_address::Response>> {
+        use zcash_address::unified::Container;
+
+        async move {
+            let unified_address_data: (zcash_address::Network, zcash_address::unified::Address) =
+                zcash_address::unified::Encoding::decode(address.clone().as_str()).map_err(
+                    |error| Error {
+                        code: ErrorCode::ServerError(0),
+                        message: error.to_string(),
+                        data: None,
+                    },
+                )?;
+
+            let mut p2pkh = String::new();
+            let mut p2sh = String::new();
+            let mut orchard = String::new();
+            let mut sapling = String::new();
+
+            for item in unified_address_data.1.items() {
+                match item {
+                    zcash_address::unified::Receiver::Orchard(_data) => {
+                        if let Ok(addr) =
+                            zcash_address::unified::Address::try_from_items(vec![item])
+                        {
+                            orchard = addr.encode(&unified_address_data.0);
+                        }
+                    }
+                    zcash_address::unified::Receiver::Sapling(data) => {
+                        if let Ok(addr) = zebra_chain::primitives::Address::try_from_sapling(
+                            unified_address_data.0,
+                            data,
+                        ) {
+                            sapling = addr.payment_address().unwrap_or_default();
+                        }
+                    }
+                    zcash_address::unified::Receiver::P2pkh(data) => {
+                        if let Ok(addr) =
+                            zebra_chain::primitives::Address::try_from_transparent_p2pkh(
+                                unified_address_data.0,
+                                data,
+                            )
+                        {
+                            p2pkh = addr.payment_address().unwrap_or_default();
+                        }
+                    }
+                    zcash_address::unified::Receiver::P2sh(data) => {
+                        if let Ok(addr) =
+                            zebra_chain::primitives::Address::try_from_transparent_p2sh(
+                                unified_address_data.0,
+                                data,
+                            )
+                        {
+                            p2sh = addr.payment_address().unwrap_or_default();
+                        }
+                    }
```

Comment:
Optional: I think we might want to split up the conversion and encoding parts of this code.

1. The conversion part looks like an implementation of:

    ```rust
    impl zcash_address::TryFromAddress for zebra_chain::primitives::Address {
        fn try_from_unified(net: Network, data: Address) -> Result<Self, ConversionError<Self::Error>> { ... }
    }
    ```

    The rest of that code is here:
    https://github.com/ZcashFoundation/zebra/blob/3f1d94dd87d0dbb85bb538faed1bbb6c10979ccd/zebra-chain/src/primitives/address.rs#L62

    If we move the conversion to that module, the error handling will be a lot easier, and the code will be shorter. And we can re-use it in other RPCs and other parts of Zebra.

2. We could move the encoding part into a `Display` impl for each type in `zebra_chain::primitives::Address::Unified`. (That's already the way it works for transparent.) That way, we can do the encoding separately, and re-use it in other RPCs.

3. I also think we might find it easier to work with unified addresses if we have methods on `zebra_chain::primitives::Address` called:
    - `transparent() -> Option<transparent::Address>`
    - `sapling() -> Option<sapling::PaymentAddress>`
    - `orchard() -> Option<orchard::Address>`

    I think we'll need them for this RPC and the other address RPC.

Reply:
This seems to have been sorted out by other changes.